### ISSUE TITLE
Auth header error when pulling Docker image to AWS instance

### ISSUE THREAD

**Issue by kramimus** [apptainer/singularity#1240]
## Version of Singularity:
```
$ singularity --version
2.4.2-master.g91881f75
```

### Expected behavior
Running 
```
$ singularity pull docker://tensorflow/tensorflow
```
on an AWS instance should pull the latest *public* tensorflow docker image into a singularity .simg

### Actual behavior
There seems to be some issue with downloading the docker image layers:
```
$ singularity pull docker://tensorflow/tensorflow
WARNING: pull for Docker Hub is not guaranteed to produce the
WARNING: same image on repeated pull. Use Singularity Registry
WARNING: (shub://) to pull exactly equivalent images.
Docker image path: index.docker.io/tensorflow/tensorflow:latest
Cache folder set to /home/ubuntu/.singularity/docker
[12/12] |===================================| 100.0% 
Importing: base Singularity environment
Importing: /home/ubuntu/.singularity/docker/sha256:054be6183d067af1af06196d7123f7dd0b67f7157a0959bd857ad73046c3be9a.tar.gz

gzip: /home/ubuntu/.singularity/docker/sha256:054be6183d067af1af06196d7123f7dd0b67f7157a0959bd857ad73046c3be9a.tar.gz: not in gzip format
tar: This does not look like a tar archive
tar: Exiting with failure status due to previous errors
Cleaning up...
ERROR: pulling container failed!
ubuntu@ip-10-23-0-47:/tmp$ cat /home/ubuntu/.singularity/docker/sha256:054be6183d067af1af06196d7123f7dd0b67f7157a0959bd857ad73046c3be9a.tar.gz
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>InvalidArgument</Code><Message>Only one auth mechanism allowed; only the X-Amz-Algorithm query parameter, Signature query string parameter or the Authorization header should be specified</Message><ArgumentName>Authorization</ArgumentName><ArgumentValue>Bearer TOKEN HERE
```
Note pulling the same tensorflow version using singularity built the same way locally (non-AWS instance,  ubuntu 16.04) builds the .simg file just fine. For some reason this is only happening on AWS instances (also ubuntu 16.04)

This was also working fine at some point on AWS instances (month or 2 ago?), only recently started failing.

### Steps to reproduce behavior

- Start AWS ubuntu 16.04 instance (c4.large in this case, although I don't think it matters, just AWS specific)
- Clone and build from singularity master
- Run `singularity pull docker://tensorflow/tensorflow`

Doing a native `docker pull tensorflow/tensorflow` on the same ubuntu AWS instance yields a working docker image (runs fine in docker).

**Comment by vsoch** [comment]
check out @jtriley work on this - he had a patch to fix it some time ago! 

https://github.com/singularityware/singularity/pull/1184

**Comment by kramimus** [comment]
Thanks, posted my results to this related issue: https://github.com/singularityware/singularity/issues/1159

**Comment by JiaweiZhuang** [comment]
Is this fixed? I have the same issue with multiple OS on AWS. Even just the "hello world" example won't work. 

On Amazon Linux (based on RHEL):
```
[ec2-user@ip-172-31-66-96 ~]$ singularity pull --name funny.simg docker://godlovedc/lolcow
WARNING: pull for Docker Hub is not guaranteed to produce the
WARNING: same image on repeated pull. Use Singularity Registry
WARNING: (shub://) to pull exactly equivalent images.
Docker image path: index.docker.io/godlovedc/lolcow:latest
Cache folder set to /home/ec2-user/.singularity/docker
Importing: base Singularity environment
Importing: /home/ec2-user/.singularity/docker/sha256:9fb6c798fa41e509b58bccc5c29654c3ff4648b608f5daa67c1aab6a7d02c118.tar.gz

gzip: /home/ec2-user/.singularity/docker/sha256:9fb6c798fa41e509b58bccc5c29654c3ff4648b608f5daa67c1aab6a7d02c118.tar.gz: not in gzip format
tar: This does not look like a tar archive
tar: Exiting with failure status due to previous errors
Cleaning up...
ERROR: pulling container failed!
```

On ubuntu 16.04:
```
ubuntu@ip-172-31-69-217:~/singularity-2.4.2$ singularity pull --name funny.simg docker://godlovedc/lolcow
WARNING: pull for Docker Hub is not guaranteed to produce the
WARNING: same image on repeated pull. Use Singularity Registry
WARNING: (shub://) to pull exactly equivalent images.
Docker image path: index.docker.io/godlovedc/lolcow:latest
Cache folder set to /home/ubuntu/.singularity/docker
Importing: base Singularity environment
Importing: /home/ubuntu/.singularity/docker/sha256:9fb6c798fa41e509b58bccc5c29654c3ff4648b608f5daa67c1aab6a7d02c118.tar.gz

gzip: /home/ubuntu/.singularity/docker/sha256:9fb6c798fa41e509b58bccc5c29654c3ff4648b608f5daa67c1aab6a7d02c118.tar.gz: not in gzip format
tar: This does not look like a tar archive
tar: Exiting with failure status due to previous errors
Cleaning up...
ERROR: pulling container failed!
```

**Comment by vsoch** [comment]
ah I've been looking for this bug! Could you also post here https://github.com/singularityhub/sregistry-cli and I'll debug for you later today? To be specific - you are running on Amazon EC2 or are you also able to produce the issue here? https://hub.docker.com/_/amazonlinux/ Thanks for posting!

**Comment by JiaweiZhuang** [comment]
@vsoch I've posted: singularityhub/sregistry-cli#72

Using Singularity inside the Amazon Linux Docker container seems fine:

```
bash-4.2# singularity pull --name funny.simg docker://godlovedc/lolcow
WARNING: pull for Docker Hub is not guaranteed to produce the
WARNING: same image on repeated pull. Use Singularity Registry
WARNING: (shub://) to pull exactly equivalent images.
Docker image path: index.docker.io/godlovedc/lolcow:latest
Cache folder set to /root/.singularity/docker
[6/6] |===================================| 100.0%
Importing: base Singularity environment
Importing: /root/.singularity/docker/sha256:9fb6c798fa41e509b58bccc5c29654c3ff4648b608f5daa67c1aab6a7d02c118.tar.gz
Importing: /root/.singularity/docker/sha256:3b61febd4aefe982e0cb9c696d415137384d1a01052b50a85aae46439e15e49a.tar.gz
Importing: /root/.singularity/docker/sha256:9d99b9777eb02b8943c0e72d7a7baec5c782f8fd976825c9d3fb48b3101aacc2.tar.gz
Importing: /root/.singularity/docker/sha256:d010c8cf75d7eb5d2504d5ffa0d19696e8d745a457dd8d28ec6dd41d3763617e.tar.gz
Importing: /root/.singularity/docker/sha256:7fac07fb303e0589b9c23e6f49d5dc1ff9d6f3c8c88cabe768b430bdb47f03a9.tar.gz
Importing: /root/.singularity/docker/sha256:8e860504ff1ee5dc7953672d128ce1e4aa4d8e3716eb39fe710b849c64b20945.tar.gz
Importing: /root/.singularity/metadata/sha256:736a219344fbca3099ce5bd1d2dbfea74b22b830bac0e85ecca812c2983390cd.tar.gz
Building Singularity image...
Singularity container built: ./funny.simg
Cleaning up...
Done. Container is at: ./funny.simg
```

However, running that image inside Docker leads to an error. But this might be a Docker problem (never tried to use Singularity inside Docker before). 
```
bash-4.2# ./funny.simg
Singularity: action-suid (U=0,P=26264)> Could not virtualize file system namespace: Operation not permitted

ERROR  : Could not virtualize file system namespace: Operation not permitted
Singularity: action-suid (U=0,P=26264)> Retval = 255

ABORT  : Retval = 255
```

**Comment by vsoch** [comment]
okay I'm debugging this for you now, on Amazon.

**Comment by vsoch** [comment]
Replicated and was (already fixed) and shown here https://github.com/singularityhub/sregistry-cli/issues/72

**Comment by JiaweiZhuang** [comment]
> Replicated and was (already fixed) and shown here singularityhub/sregistry-cli#72

Thanks, so let me confirm the solution: "Whenever on EC2, one should use sregistry instead of the original singularity command to pull Docker containers"?

**Comment by vsoch** [comment]
As long as the bug exists in the singularity core (hopefully already fixed in development branch) this is the best way to get your image.

**Comment by kramimus** [comment]
The branch mentioned in this PR fixes the pull issue on EC2 for me

https://github.com/singularityware/singularity/issues/1159

**Comment by dtrudg** [comment]
This issue should be fixed on 2.4.5 thanks to PR #1387

**Comment by soichih** [comment]
I was hit by this issue. Ubuntu 18 still distributes 2.4.2 right at this moment.. Do you know when 2.4.5 will be released via the official channel?
